src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -278,12 +271,58 @@ private void CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
             }
         }
 
-        private void CreateEmbeddedAttributeItselfIfNeeded(DiagnosticBag diagnostics)
+        protected override bool InjectedSymbolsAreFrozen => !_lazyInjectedTypes.IsDefault;
+
+        /// <summary>
+        /// Get injected types that are needed (ie. were used) and report the diagnostics they held onto (only once).
+        /// </summary>
+        protected sealed override ImmutableArray<NamedTypeSymbol> GetInjectedTypes(DiagnosticBag diagnostics)
         {
-            CreateEmbeddedAttributeIfNeeded(
-                ref _lazyEmbeddedAttribute,
-                diagnostics,
-                AttributeDescription.CodeAnalysisEmbeddedAttribute);
+            if (!_lazyInjectedTypes.IsDefault)
+            {
+                return _lazyInjectedTypes;
+            }
+
+            if (_needsNonNullTypesAttribute)
+            {
+                EnsureEmbeddedAttributeExists();
+            }
+
+            ArrayBuilder<NamedTypeSymbol> builder = null;
+            if (_needsEmbeddedAttribute)
+            {
+                addInjectedAttribute(WellKnownType.Microsoft_CodeAnalysis_EmbeddedAttribute, canUseFromSource: false);
+            }
+
+            if (_needsNonNullTypesAttribute)
+            {
+                addInjectedAttribute(WellKnownType.System_Runtime_CompilerServices_NonNullTypesAttribute, canUseFromSource: true);
+            }
+
+            var result = builder is null ? ImmutableArray<NamedTypeSymbol>.Empty : builder.ToImmutableAndFree();
+            ImmutableInterlocked.InterlockedInitialize(ref _lazyInjectedTypes, result);
+            return _lazyInjectedTypes;
+
+            void addInjectedAttribute(WellKnownType wellKnownType, bool canUseFromSource)
+            {
+                NamedTypeSymbol attribute = Compilation.GetWellKnownType(wellKnownType);
```

Comment:
>Compilation.GetWellKnownType [](start = 44, length = 28)

Can this return a symbol from an added module?

Reply:
Added a test (crash in emit). Thanks

Filed a follow-up issue for this and also duplicate diagnostic issues: [https://github.com/dotnet/roslyn/issues/29732](https://github.com/dotnet/roslyn/issues/29732)

---
In reply to: [216098979](https://github.com/dotnet/roslyn/pull/29180#discussion_r216098979) [](ancestors = 216098979)